### PR TITLE
Use fence to apply stricter ordering for optimisic read of state.

### PR DESCRIPTION
**Pull request by PaulSandoz**
See comment and http://gee.cs.oswego.edu/dl/html/j9mm.html (and the older http://gee.cs.oswego.edu/dl/jmm/cookbook.html) for more details.

### CODE REVIEW

File: hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java [hunk 1]
```diff
@@ -373,6 +374,31 @@ HollowObjectTypeDataElements currentDataElements() {
     }
 
     private boolean readWasUnsafe(HollowObjectTypeDataElements data) {
+        // Use a load (acquire) fence to constrain the compiler reordering prior plain loads so
+        // that they cannot "float down" below the volatile load of currentDataVolatile.
+        // This ensures data is checked against currentData *after* optimistic calculations
+        // have been performed on data.
+        //
+        // Note: the Java Memory Model allows for the reordering plain loads and stores
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java]: I think you meant either "for reordering" or "for the reordering of"

File: hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java [hunk 2]
```diff
@@ -373,6 +374,31 @@ HollowObjectTypeDataElements currentDataElements() {
...
+        // from above (float down the acquire) or below (float above the release) but existing
+        // loads and stores may not exit (a "lock roach motel" and why there is almost universal
+        // misunderstanding of, and many misguided attempts to optimize, the infamous double
+        // checked locking idiom).
+        //
+        // Note: the fence provides stronger ordering guarantees than a corresponding non-plain
+        // load or store since the former affects all prior or subsequent loads and stores,
+        // whereas the latter is scoped to the particular load or store.
+        //
+        // For more details see http://gee.cs.oswego.edu/dl/html/j9mm.html
```
Comment (ghost) [inline hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java:400]: This comment read clearly to me. 👍

**Review by ghost — APPROVED** [review]

File: hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java (hunk 1 above)
Comment (PaulSandoz) [reply hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java]: Yes, i shall fix it.

File: hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java (hunk 2 above)
Comment (PaulSandoz) [reply hollow/src/main/java/com/netflix/hollow/core/read/engine/object/HollowObjectTypeReadStateShard.java:400]: Thanks!

**Review by dkoszewnik — APPROVED** [review]

**Review by dkoszewnik — APPROVED** [review]

**Pull request merged**